pkg/statistics/builder.go

```diff
@@ -287,97 +294,97 @@ func BuildHistAndTopN(
+		if err != nil {
+			return true
 		}
-		// case 2-2, now topn is full, and the "current" count is less than the least count in the topn: no need to insert the "current"
-		if len(topNList) >= numTopN && uint64(curCnt) <= topNList[len(topNList)-1].Count {
-			cur, curCnt = sampleBytes, 1
-			continue
+		return cmp < 0
+	})
+	sampleTree := btree.NewG(8, func(a, b *sortItem) bool {
+		var cmp int
+		cmp, err = a.Sample.Value.Compare(sc.TypeCtx(), &b.Sample.Value, collate.GetBinaryCollator())
+		if err != nil {
+			return true
 		}
-		// case 2-3, now topn is not full, or the "current" count is larger than the least count in the topn: need to find a slot to insert the "current"
-		j := len(topNList)
-		for ; j > 0; j-- {
-			if uint64(curCnt) < topNList[j-1].Count {
-				break
-			}
+		return cmp < 0
+	})
+	for _, value := range sMap {
+		topNbtree.ReplaceOrInsert(value)
+		if err != nil {
+			return nil, nil, errors.Trace(err)
 		}
-		topNList = append(topNList, TopNMeta{})
-		copy(topNList[j+1:], topNList[j:])
-		topNList[j] = TopNMeta{Encoded: cur, Count: uint64(curCnt)}
-		if len(topNList) > numTopN {
-			topNList = topNList[:numTopN]
+		for topNbtree.Len() > numTopN {
+			topNbtree.DeleteMin()
+		}
+		sampleTree.ReplaceOrInsert(value)
+		if err != nil {
+			return nil, nil, errors.Trace(err)
 		}
-		cur, curCnt = sampleBytes, 1
 	}
-
 	// Calc the correlation of the column between the handle column.
 	if isColumn {
 		hg.Correlation = calcCorrelation(sampleNum, corrXYSum)
 	}
-
+	var topNList []TopNMeta
 	// Handle the counting for the last value. Basically equal to the case 2 above.
 	// now topn is empty: append the "current" count directly
-	if len(topNList) == 0 {
-		topNList = append(topNList, TopNMeta{Encoded: cur, Count: uint64(curCnt)})
-	} else if len(topNList) < numTopN || uint64(curCnt) > topNList[len(topNList)-1].Count {
-		// now topn is not full, or the "current" count is larger than the least count in the topn: need to find a slot to insert the "current"
-		j := len(topNList)
-		for ; j > 0; j-- {
-			if uint64(curCnt) < topNList[j-1].Count {
-				break
-			}
-		}
-		topNList = append(topNList, TopNMeta{})
-		copy(topNList[j+1:], topNList[j:])
-		topNList[j] = TopNMeta{Encoded: cur, Count: uint64(curCnt)}
-		if len(topNList) > numTopN {
-			topNList = topNList[:numTopN]
+	topNbtree.Descend(func(i *sortItem) bool {
+		var encoded []byte
+		encoded, err = getComparedBytes(i.Sample.Value)
+		if err != nil {
+			return false
 		}
+		topNList = append(topNList, TopNMeta{Count: uint64(i.Count), Encoded: encoded})
+		return true
+	})
+	if err != nil {
+		return nil, nil, errors.Trace(err)
 	}
 
 	topNList = pruneTopNItem(topNList, ndv, nullCount, sampleNum, count)
-
+	samples = make([]*SampleItem, 0, sampleTree.Len())
+	for sampleTree.Len() != 0 {
+		item, _ := sampleTree.DeleteMin()
+		samples = append(samples, item.Sample)
+	}
 	// Step2: exclude topn from samples
 	for i := int64(0); i < int64(len(samples)); i++ {
```

Comment:
Key Changes:
Introduced a Map for Efficient Lookup: A map topNMap is created to store the topNList items. This allows for O(1) lookup time to check if a sample is in the topN list.

Eliminated the Inner Loop: The inner loop that iterated over topNList for each sample is removed. Instead, the map is used for checking the presence of a sample in the topN list.

Preserving Samples Not in TopN: A new slice newSamples is created to hold samples that are not in the topN list. This avoids modifying the original samples slice while iterating over it.

This optimization should significantly reduce the time complexity, particularly when dealing with a large number of samples and a sizable topN list. However, it's important to note that optimizations should always be validated with real-world data and use cases to ensure they provide the expected benefits.


// Part of BuildHistAndTopN function with optimization
topNMap := make(map[string]uint64) // Create a map to hold topN items for quick lookup
for _, topN := range topNList {
    topNMap[string(topN.Encoded)] = topN.Count
}

newSamples := make([]*SampleItem, 0, len(samples))
for _, sample := range samples {
    sampleBytes, err := getComparedBytes(sample.Value)
    if err != nil {
        return nil, nil, errors.Trace(err)
    }

    if count, exists := topNMap[string(sampleBytes)]; exists {
        // If found in topN, skip the next 'count' samples
        continue
    }
    newSamples = append(newSamples, sample) // Keep the sample if not in topN
}

samples = newSamples // Replace old samples with the filtered ones

Reply:
I think If NDV is high, we can merge sampling by hashmap. so it can avoid too much compare when to sort. I need to add more benchmark to verify it. but if NDV is low, it is unavailable.